Subject: softirq: Split handling function
From: Thomas Gleixner <tglx@linutronix.de>
Date: Thu, 04 Oct 2012 15:33:53 +0100
Origin: https://www.kernel.org/pub/linux/kernel/projects/rt/3.14/patches-3.14.10-rt7.tar.xz

Split out the inner handling function, so RT can reuse it.

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/softirq.c |   34 +++++++++++++++++++++-------------
 1 file changed, 21 insertions(+), 13 deletions(-)

--- a/kernel/softirq.c
+++ b/kernel/softirq.c
@@ -141,10 +141,29 @@ static void wakeup_softirqd(void)
 		wake_up_process(tsk);
 }
 
+static void handle_softirq(unsigned int vec_nr, int cpu, int need_rcu_bh_qs)
+{
+	struct softirq_action *h = softirq_vec + vec_nr;
+	unsigned int prev_count = preempt_count();
+
+	kstat_incr_softirqs_this_cpu(vec_nr);
+
+	trace_softirq_entry(vec_nr);
+	h->action(h);
+	trace_softirq_exit(vec_nr);
+	if (unlikely(prev_count != preempt_count())) {
+		pr_err("huh, entered softirq %u %s %p with preempt_count %08x, exited with %08x?\n",
+				vec_nr, softirq_to_name[vec_nr], h->action,
+				prev_count, preempt_count());
+		preempt_count_set(prev_count);
+	}
+	if (need_rcu_bh_qs)
+		rcu_bh_qs(cpu);
+}
+
 static void handle_pending_softirqs(u32 pending, int cpu, int need_rcu_bh_qs)
 {
 	struct softirq_action *h = softirq_vec;
-	unsigned int prev_count = preempt_count();
 	int softirq_bit;
 
 	local_irq_enable();
@@ -156,19 +175,8 @@ static void handle_pending_softirqs(u32
 		h += softirq_bit - 1;
 		vec_nr = h - softirq_vec;
 
-		kstat_incr_softirqs_this_cpu(vec_nr);
+		handle_softirq(vec_nr, cpu, need_rcu_bh_qs);
 
-		trace_softirq_entry(vec_nr);
-		h->action(h);
-		trace_softirq_exit(vec_nr);
-		if (unlikely(prev_count != preempt_count())) {
-			pr_err("huh, entered softirq %u %s %p with preempt_count %08x, exited with %08x?\n",
-			       vec_nr, softirq_to_name[vec_nr], h->action,
-			       prev_count, preempt_count());
-			preempt_count_set(prev_count);
-		}
-		if (need_rcu_bh_qs)
-			rcu_bh_qs(cpu);
 		h++;
 		pending >>= softirq_bit;
 	}
